aiidalab_widgets_base/structures.py

```diff
@@ -114,68 +125,84 @@ def store_structure(self, label=None, description=None):
             return
         if self.structure_node.is_stored:
-            print("Already stored in AiiDA: " + repr(self.structure_node) + " skipping..")
+            self.output.value = "Already stored in AiiDA [{}], skipping...".format(self.structure_node)
             return
         if label:
             self.structure_node.label = label
         if description:
             self.structure_node.description = description
         self.structure_node.store()
-        print("Stored in AiiDA: " + repr(self.structure_node))
+        self.output.value = "Stored in AiiDA [{}]".format(self.structure_node)
 
-    @observe('node_class')
-    def _change_structure_node(self, change):
-        if change['new'] is not None:
-            self.structure = self._inserted_structure
+    @default('node_class')
+    def _default_node_class(self):
+        return 'StructureData'
 
-    @validate('structure')
-    def _valid_structure(self, change):
+    @observe('node_class')
+    def _change_structure_node(self, _=None):
+        self.structure_node = self._inserted_structure
+
+    @default('structure')
+    def _default_structure(self):
+        return None
+
+    @validate('structure_node')
+    def _validate_structure_node(self, proposal):
+        """Validate structure_node trait."""
+        structure_node = proposal['value']
+        self._inserted_structure = structure_node
+
+        StructureNode = DataFactory(self.SUPPORTED_DATA_FORMATS[self.node_class])  # pylint: disable=invalid-name
+        if isinstance(structure_node, StructureNode):
+            return structure_node
+        return StructureNode(ase=structure_node.get_ase())
+
+    @observe('structure_node')
+    def _update_structure(self, change):
         """Returns ASE atoms object and sets structure_node trait."""
+        structure_node = change['new']
 
-        self._inserted_structure = change['value']
-        structure_ase = None
-        structure_node = None
+        if structure_node is None:
+            with self.hold_trait_notifications():
+                self.structure = None
+                self.btn_store.disabled = True
+                structure_ase = None
+        else:
+            self.btn_store.disabled = False
+            structure_ase = structure_node.get_ase()
 
-        # Chosing type for the structure node.
-        StructureNode = DataFactory(self.DATA_FORMATS[self.node_class])  # pylint: disable=invalid-name
+        if self.structure != structure_ase:
+            with self.hold_trait_notifications():
+                self.structure = structure_ase
 
-        # If no structure is provided, the rest of the code can be skipped
-        if self._inserted_structure is None:
-            return None
+    @observe('structure')
+    def _update_structure_node(self, change=None):
+        """This function enables/disables `btn_store` widget if structure is provided/set to None.
+        Also, the function sets `structure_node` trait to the selected node type."""
```

Comment:
This doc-string is non standard. I'm surprised the linter didn't complain.

Reply:
How should it look like? I don't know what is the problem here.

Reply:
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Reply:
fixed.